pkg/container-utils/containerd/containerd.go

```diff
@@ -15,31 +15,213 @@
+	"github.com/containerd/containerd"
+	tasks "github.com/containerd/containerd/api/services/tasks/v1"
+	"github.com/containerd/containerd/api/types/task"
+
 	runtimeclient "github.com/inspektor-gadget/inspektor-gadget/pkg/container-utils/runtime-client"
 )
 
 const (
 	DefaultTimeout = 2 * time.Second
+
+	LabelK8sContainerName         = "io.kubernetes.container.name"
+	LabelK8sContainerdKind        = "io.cri-containerd.kind"
+	LabelK8sContainerdKindSandbox = "sandbox"
 )
 
 type ContainerdClient struct {
-	criclient.CRIClient
+	client *containerd.Client
 }
 
 func NewContainerdClient(socketPath string) (runtimeclient.ContainerRuntimeClient, error) {
 	if socketPath == "" {
 		socketPath = runtimeclient.ContainerdDefaultSocketPath
 	}
 
-	criClient, err := criclient.NewCRIClient(runtimeclient.ContainerdName, socketPath, DefaultTimeout)
+	client, err := containerd.New(socketPath,
+		containerd.WithTimeout(DefaultTimeout),
+		containerd.WithDefaultNamespace("k8s.io"),
+	)
 	if err != nil {
 		return nil, err
 	}
 
 	return &ContainerdClient{
-		CRIClient: criClient,
+		client: client,
+	}, nil
+}
+
+func (c *ContainerdClient) Close() error {
+	if c.client != nil {
+		return c.client.Close()
+	}
+	return nil
+}
+
+func (c *ContainerdClient) isSandboxContainer(container containerd.Container) bool {
+	labels, err := container.Labels(context.TODO())
+	if err != nil {
+		return false
+	}
+
+	if kind, ok := labels[LabelK8sContainerdKind]; ok {
+		return kind == LabelK8sContainerdKindSandbox
+	}
+
+	return false
+}
+
+func (c *ContainerdClient) GetContainers() ([]*runtimeclient.ContainerData, error) {
+	taskResponse, err := c.client.TaskService().List(context.TODO(), &tasks.ListTasksRequest{})
+	if err != nil {
+		return nil, err
+	}
+
+	ret := make([]*runtimeclient.ContainerData, 0, len(taskResponse.Tasks))
+	for _, task := range taskResponse.Tasks {
+
+		container, err := c.getContainerdContainer(task.ID)
+		if err != nil {
+			return nil, err
+		}
+
+		if c.isSandboxContainer(container) {
+			continue
+		}
+
+		containerData, err := c.containerdTaskAndContainerToContainerData(task, container)
+		if err != nil {
+			return nil, err
+		}
+
+		ret = append(ret, containerData)
+	}
+
+	return ret, nil
+}
+
+func (c *ContainerdClient) GetContainer(containerID string) (*runtimeclient.ContainerData, error) {
+	response, err := c.client.TaskService().Get(context.TODO(), &tasks.GetRequest{
+		ContainerID: containerID,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	containerData, err := c.ContainerdTaskToContainerData(response.Process)
+	if err != nil {
+		return nil, err
+	}
+	return containerData, nil
+}
+
+func (c *ContainerdClient) GetContainerDetails(containerID string) (*runtimeclient.ContainerDetailsData, error) {
+	response, err := c.client.TaskService().Get(context.TODO(), &tasks.GetRequest{
+		ContainerID: containerID,
+	})
+	if err != nil {
+		return nil, err
+	}
+	proc := response.Process
+
+	container, err := c.getContainerdContainer(proc.ID)
```

Comment:
I also expected it to be `ContainerID` but that field is empty for containers i created via `ctr` or `kubernetes`.

I can't recall what that field was used for. But it would be useful to know what it represents or had represented in the past?

```bash
$ sudo ctr c create docker.io/library/redis:alpine redis_container_id
$ sudo ctr c ls
CONTAINER             IMAGE                              RUNTIME                  
redis_container_id    docker.io/library/redis:alpine     io.containerd.runc.v2
$ sudo ctr t start -d redis_container_id
$ sudo ctr t ls
TASK                  PID       STATUS    
redis_container_id    620763    RUNNING
```

Source for `ctr t ls` is https://github.com/containerd/containerd/blob/main/cmd/ctr/commands/tasks/list.go#L62 and that is reads the field `ID`

Reply:
Mmmm, I think it is a problem of how `ctr` show things because I can see the ID behaves as we expect for Kubernetes containers using `nerdctl`:

```bash
$ sudo ./local-gadget list-containers  -v -w -r containerd -c mypod
# Create a pod in another terminal, e.g. kubectl run --rm --image=busybox mypod -it --restart='Never' -- sh
DEBU[0041] runc fanotify: add container "22f48fc5e4d40051b2b3cb11bd39ce109591e6c18e235a70daf7a54eac1ff32c"
DEBU[0041] Runtime enricher (containerd): failed to get container: listing task for container "22f48fc5e4d40051b2b3cb11bd39ce109591e6c18e235a70daf7a54eac1ff32c": rpc error: code = NotFound desc = task 22f48fc5e4d40051b2b3cb11bd39ce109591e6c18e235a70daf7a54eac1ff32c not found
DEBU[0042] runc fanotify: add container "4d8e614ec3ce2a8a571b94ffd1e13697a7d7c52133bcb634e7f3a8c271b7eb89"
DEBU[0042] Runtime enricher (containerd): failed to get container: listing task for container "4d8e614ec3ce2a8a571b94ffd1e13697a7d7c52133bcb634e7f3a8c271b7eb89": rpc error: code = NotFound desc = task 4d8e614ec3ce2a8a571b94ffd1e13697a7d7c52133bcb634e7f3a8c271b7eb89 not found
```

The ID reported by runc-fanotify, which is the one that will be used to call `GetContainer(s)`, is the same that `nerdctl` uses to identify the containers:

```bash
$ sudo nerdctl -n k8s.io ps --filter id=4d8e614ec3ce2a8a571b94ffd1e13697a7d7c52133bcb634e7f3a8c271b7eb89
CONTAINER ID    IMAGE                               COMMAND    CREATED          STATUS    PORTS    NAMES
4d8e614ec3ce    docker.io/library/busybox:latest    "sh"       2 minutes ago    Up                 k8s://default/mypod/mypod
```

Reply:
I think the `ContainerId` field of [Task.Process](https://pkg.go.dev/github.com/containerd/containerd/api/types/task#Process) just could be an old field that is now unused?

When creating a container through `nerctl` i get the following values for `Task.Process.ID` and `Task.Process.ContaienerID`:
```bash
$ sudo nerdctl run -d --name nginxtest -p 8080:80 library/nginx:1.22.1-alpine 
docker.io/library/nginx:1.22.1-alpine:                                            resolved       |++++++++++++++++++++++++++++++++++++++| 
index-sha256:2366ede62d2e26a20f7ce7d0294694fe52b166107fd346894e4658dfb5273f9c:    done
....
layer-sha256:71a3f979d23b00e5731837eef4583847c2d56e35fded467d0a79d5853373b533:    done           |++++++++++++++++++++++++++++++++++++++| 
elapsed: 3.9 s                                                                    total:  9.8 Mi (2.5 MiB/s)                                       
9a9531aefbf7e18b873db8d852f43e0d83bbed9b38e658c7de63289a61d421bb

$ sudo nerdctl ps                                                            
CONTAINER ID    IMAGE                                    COMMAND                   CREATED           STATUS    PORTS                   NAMES
9a9531aefbf7    docker.io/library/nginx:1.22.1-alpine    "/docker-entrypoint.…"    25 seconds ago    Up        0.0.0.0:8080->80/tcp    nginxtest                  

$ sudo ./local-gadget list-containers -v --runtimes=containerd   (LOCAL DEBUG OUTPUT)
....
Container proc id: 9a9531aefbf7e18b873db8d852f43e0d83bbed9b38e658c7de63289a61d421bb
Container proc container id: 
....
``` 

Containers that kubernetes created:
```bash
$ sudo ./local-gadget list-containers -v --runtimes=containerd   (LOCAL DEBUG OUTPUT in Minikube)
....
Container proc id: f806e6391201048ff67266f94e608eaabee9d0ec34b4c60c7373a640e2b9b0c2
Container proc container id: 
....
containerd f806e6391201048ff67266f94e608eaabee9d0ec34b4c60c7373a640e2b9b0c2 kindnet-cni
```

[Task.Process.ContainerID](https://pkg.go.dev/github.com/containerd/containerd/api/types/task#Process) is somehow always empty. And [Task.Process.ID](https://pkg.go.dev/github.com/containerd/containerd/api/types/task#Process) seems to be same as [Container.ID()](https://pkg.go.dev/github.com/containerd/containerd#Container)

Reply:
Another possibility could be the following:
Instead of asking the API to get all current running Tasks we could list all containers.
Then we could query the container if a Task is attached to the container.

This way we don't need to handle instances of `Task.Process`. But there is the possibility that we are calling more API calls than in this current implementation.

Reply:
> Instead of asking the API to get all current running Tasks we could list all containers.
Then we could query the container if a Task is attached to the container.

I think this is a clearer approach.

Reply:
Implemented that approach